R/plot.R

```diff
@@ -2,62 +2,145 @@
-  ggplot2::ggplot(melted, ggplot2::aes(x=value, fill=variable)) +
+plot_density<- function(...) {
+  dfs <- list(...)
+  joined <- join_dfs(dfs)
+  melted <- reshape2::melt(joined, variable.name="method", id=c("id"))
+  ggplot2::ggplot(melted, ggplot2::aes(x=value, fill=method)) +
     ggplot2::geom_density(alpha=0.5)
 }
 
 #' Produce an ROC curve which plots a given method's sensitivity/specificity with respect
 #' a given poverty threshold.
-plot_roc <- function(kfold_results, threshold) {
-  response <- kfold_results$true < threshold
-  roc <- pROC::roc(response, kfold_results$predicted)
-  plot(roc)
-}
-
-
-get_cumulative <- function(kfold_results, threshold) {
+plot_roc <- function(THRESHOLD, ...) {
+  dfs <- list(...)
+  joined <- join_dfs(dfs)
+  joined$response <- joined$true < THRESHOLD
+  joined$id <- NULL
+  joined$true <- NULL
   
+  roc_formula <- as.formula(paste("response ~", paste(names(dfs), collapse="+")))
+  rocs <- pROC::roc(roc_formula, data=joined, plot=FALSE)
+  if(length(dfs) == 1) {
+    rocs <- list(rocs)
+    names(rocs) <- names(dfs)
+  }
+  roc_to_df <- function(name, roc) {
+    data.frame(sensitivity=roc$sensitivities, specificity=roc$specificities, method=name)
+  }
+  roc_dfs <- mapply(roc_to_df, names(rocs), rocs, SIMPLIFY=FALSE)
+  roc_df <- do.call("rbind", roc_dfs)
+  ggplot2::ggplot(roc_df, ggplot2::aes(x=specificity, y=sensitivity, color=method)) +
+    ggplot2::geom_step() +
+    ggplot2::scale_x_reverse() +
+    ggplot2::geom_abline(intercept=1, slope=1, alpha=0.5) 
+  # TODO display auc
 }
 
 
 #' If we target N people, what fraction of the true poor would receive funds?
 #' True Positives / Total Positives
-plot_accuracy <- function(kfold_results, threshold, point_count=20) {
-  N <- nrow(kfold_results)
-  plot_points <- seq(1, N, length=point_count)
-  
-  ranked <- kfold_results[order(kfold_results$predicted), ]
-  ranked$response <- ranked$true < threshold
-  ranked$coverage <- cumsum(ranked$response) / sum(ranked$response)
+plot_accuracy <- function(THRESHOLD, ..., POINT_COUNT=20) {
+  dfs <- list(...)
+  joined <- join_dfs(dfs)
+  N <- nrow(joined)
+  plot_points <- seq(1, N, length=POINT_COUNT)
+  joined$response <- joined$true < THRESHOLD
+  true_poor <- sum(joined$response)
+
+  get_coverage <- function(method, df) {
+    cumsum(df[order(df[, method]), "response"]) / true_poor
+  }
+  ranked <- data.frame(mapply(get_coverage, names(dfs), list(joined), SIMPLIFY=FALSE))
   
-  cut <- ranked[plot_points, ]
+  cut <- ranked[plot_points, ,drop=FALSE]
   cut$percent_population_included <- plot_points / N
-  ggplot2::ggplot(cut, ggplot2::aes(x=percent_population_included, y=coverage)) +
-    ggplot2::geom_step()
+  melted <- reshape2::melt(
+    cut, 
+    value.name="coverage",
+    variable.name="method",
+    id="percent_population_included")
+  ggplot2::ggplot(melted, ggplot2::aes(x=percent_population_included, y=coverage, color=method)) +
+    ggplot2::geom_step() +
+    ggplot2::geom_point()
 }
 
```

Comment:
It would be good here to add a line showing the outcome when we cover F_true(THRESHOLD)% (F_true being the CDF of true consumption) of the population, and possibly another line showing the outcome when we cutoff at THRESHOLD. We could have this as an optional input to the function, e.g. DISPLAY_CUTOFFS = FALSE